elastic/synthetics
# feat: support pushing lightweight monitors

+ fix #542 
+ Adds support for pushing lightweight monitors via `.(yml|yaml)` files pattern to Kibana endpoint when invoked using `npx @elastic/synthetics push` command. 
+ Works by checking all the yml files inside the synthetics project and creates a Monitor object that matches with browser monitors to be able to push to the Kibana endpoint. Also performs duplication and error reporting as we extract the line number and other information to go back to the source location. 
+ Supports only `@every <1m>` syntax for scheduling lightweight monitors, Throws Error when it sees other CRON format for schedules. 
+ Uses location information from two places
  - Default locations from Synthetics.config.ts when project is initialized using `npx @elastic/synthetics init` command
  - Allows overriding for individual monitors via `locations` and `private_locations` fields in Heartbeat yml files.
+ Added couple of validations 
    - Ignores pushing `browser` and `disabled` monitors
    - Aborts if monitor `id | name` is not provided. 


### Testing
- Build the local project using - `npm run build`
- Run `npm link`
- Setup a brand new Synthetics project - node dist/cli.js init [optional-dir]
- Navigate to the new directory created with `init` and run `npm link @elastic/synthetics`
- Follow along and check if project settings have been initialized properly
- Create Heartbeat yaml files, copy over existing ones or use the below (some additional heartbeat yaml is included in the linked Kibana PR)
```yml
- type: http
  enabled: true
  id: my-monitor
  name: My Monitor
  urls: ["http://localhost:9200"]
  schedule: @every 10m
  locations: ["singapore"]
  private_locations: ["private-1"]
  timeout: 16s
  check.request:
    method: POST
    headers:
      'Content-Type': 'application/x-www-form-urlencoded'
  check.response:
    status: [200]
    body:
      - Saved
      - saved
```
- Validate by pushing with auth specified `SYNTHETICS_API_KEY=<key> npm run push`.
